torchtune/models/llama4/_component_builders.py

```diff
@@ -346,3 +347,772 @@ def llama4_mlp(dim: int, hidden_dim: int, quantize_base: bool = False) -> FeedFo
+                attn_dropout=attn_dropout,
+                lora_rank=lora_rank,
+                lora_alpha=lora_alpha,
+                lora_dropout=lora_dropout,
+                use_dora=use_dora,
+                quantize_base=quantize_base,
+            )
+        else:
+            self_attn = MultiHeadAttention(
+                embed_dim=embed_dim,
+                num_heads=num_heads,
+                num_kv_heads=num_kv_heads,
+                head_dim=head_dim,
+                q_proj=nn.Linear(embed_dim, num_heads * head_dim, bias=False),
+                k_proj=nn.Linear(embed_dim, num_kv_heads * head_dim, bias=False),
+                v_proj=nn.Linear(embed_dim, num_kv_heads * head_dim, bias=False),
+                output_proj=nn.Linear(embed_dim, embed_dim, bias=False),
+                pos_embeddings=pos_embeddings,
+                q_norm=q_norm,
+                k_norm=k_norm,
+                max_seq_len=max_seq_len,
+                attn_dropout=attn_dropout,
+            )
+
+        is_moe = moe_every_n_layers is None or (i + 1) % moe_every_n_layers == 0
+        if is_moe:
+            if apply_lora_to_mlp and decoder_lora:
+                mlp_layer = lora_llama4_moe(
+                    dim=embed_dim,
+                    hidden_dim=hidden_dim,
+                    lora_rank=lora_rank,
+                    lora_alpha=lora_alpha,
+                    lora_dropout=lora_dropout,
+                    num_experts=num_experts,
+                    experts_per_token=experts_per_token,
+                    use_shared_expert=use_shared_expert,
+                )
+            else:
+                mlp_layer = llama4_moe(
+                    dim=embed_dim,
+                    hidden_dim=hidden_dim,
+                    num_experts=num_experts,
+                    experts_per_token=experts_per_token,
+                    use_shared_expert=use_shared_expert,
+                )
+        else:
+            if apply_lora_to_mlp and decoder_lora:
+                mlp_layer = lora_llama4_mlp(
+                    dim=embed_dim,
+                    hidden_dim=hidden_dim,
+                    lora_rank=lora_rank,
+                    lora_alpha=lora_alpha,
+                    quantize_base=quantize_base,
+                    lora_dropout=lora_dropout,
+                    use_dora=use_dora,
+                )
+            else:
+                mlp_layer = llama3_mlp(
+                    dim=embed_dim, hidden_dim=hidden_dim, quantize_base=quantize_base
+                )
+
+        layer = TransformerSelfAttentionLayer(
+            attn=self_attn,
+            mlp=mlp_layer,
+            sa_norm=RMSNorm(dim=embed_dim, eps=norm_eps),
+            mlp_norm=RMSNorm(dim=embed_dim, eps=norm_eps),
+            mask_mod=mask_mod,
+        )
+        layers.append(layer)
+
+    layers = nn.ModuleList(layers)
+    tok_embeddings = nn.Embedding(vocab_size, embed_dim)
+    # TODO: quantize_base is not applied to final output_proj currently.
```

Comment:
Why not? Could be quite large

Reply:
this is an ancient copypasta from all our other builders. previously it resulted in some really large intermediates during quant. i have not revisited recently